packages/next/src/client/components/client-page.tsx

```diff
@@ -1,30 +1,107 @@
+    // appropriate context. We wrap differently in prerendering vs rendering
+    const store = staticGenerationAsyncStorage.getStore()
+    if (!store) {
+      throw new InvariantError(
+        'Expected staticGenerationStore to exist when handling searchParams in a client Page.'
+      )
+    }
+
+    if (store.isStaticGeneration) {
+      // We are in a prerender context
+      // We need to recover the underlying searchParams from the server
+      const { reifyClientPrerenderSearchParams } =
+        require('../../server/request/search-params') as typeof import('../../server/request/search-params')
+      clientSearchParams = reifyClientPrerenderSearchParams(store)
+
+      const { reifyClientPrerenderParams } =
+        require('../../server/request/params') as typeof import('../../server/request/params')
+
+      const fallbackParamNames = fp
+      if (fallbackParamNames) {
+        // the params promise won't ever resolve because we're in a PPR fallback prerender
+        // We instead expect there to be an underlying params object which gives the client
+        // access to the params that can be used to reify an aborting or interrupting params
+        // object.
+        const underlyingParams = up
+        if (!underlyingParams) {
+          throw new InvariantError(
+            `In route ${store.route}, expected params to be available to client Segment but none were found.`
+          )
+        }
+        clientParams = reifyClientPrerenderParams(
+          underlyingParams,
+          fallbackParamNames,
+          store
+        )
+      } else {
+        const underlyingParams = use(p)
+        clientParams = reifyClientPrerenderParams(
+          underlyingParams,
+          undefined,
+          store
+        )
+      }
+    } else {
+      // We are in a dynamic context and need to unwrap the underlying searchParams
+
+      // We can't type that searchParams is passed but since we control both the definition
+      // of this component and the usage of it we can assume it
+      const underlyingSearchParams = use(sp)
+      const underlyingParams = use(p)
+
+      const { reifyClientRenderSearchParams } =
+        require('../../server/request/search-params') as typeof import('../../server/request/search-params')
+      clientSearchParams = reifyClientRenderSearchParams(
+        underlyingSearchParams,
+        store
+      )
+      const { reifyClientRenderParams } =
+        require('../../server/request/params') as typeof import('../../server/request/params')
+      clientParams = reifyClientRenderParams(underlyingParams, store)
+    }
+
+    return <C params={clientParams} searchParams={clientSearchParams} />
+  } else {
+    const underlyingSearchParams = use(sp)
+    const underlyingParams = use(p)
+
+    const { reifyClientRenderSearchParams } =
+      require('../../server/request/search-params.browser') as typeof import('../../server/request/search-params.browser')
```

Comment:
is browser implementation to be used as expected when static generation is false but still in SSR? sounds like `.client`?

Reply:
the SSR one uses some things like AsyncLocalStorage and the browser one doesn't. They're largely the same but this helps ensure no server code gets bundled into the browser builds